### ISSUE TITLE
Reference a "reference table" from within a propagated function

### ISSUE THREAD

**Issue by mkurz** [citusdata/citus#2126]
```sql
CREATE TABLE my_table (
    my_column TEXT
);
select create_reference_table('my_table');
```
On the worker `my_table_102137` will be created by citus.

Now I want to propogate a function to the worker that references `my_table`:
```sql
SELECT run_command_on_workers($cmd$
    CREATE OR REPLACE FUNCTION my_func() RETURNS TEXT AS $$
        SELECT my_column FROM my_table;
    $$ LANGUAGE SQL;
$cmd$);
```
However I get an error:
```
                         run_command_on_workers                          
-------------------------------------------------------------------------
 (citus_worker_1,5432,f,"ERROR:  relation ""key_vault"" does not exist")
```
I suppose that's because I somehow should reference table `my_table_102137` from within my function.

But how do I get the name of the `my_table` shard on the worker?

**Comment by onderkalaci** [comment]
I think it is documented here: https://docs.citusdata.com/en/v7.3/develop/reference_propagation.html

An example:
```SQL
SELECT run_command_on_placements('users_table_ref'::regclass,
    $cmd$ CREATE OR REPLACE FUNCTION my_func() RETURNS TEXT AS $$
        SELECT user_id::text FROM %s;
    $$ LANGUAGE SQL;
$cmd$);
```

**Comment by mkurz** [comment]
@onderkalaci Thanks for you reply!
However I don't fully understand - the docs don't say anything about the `::regclass` cast you did.
Can you explain what your example exactly does? It wouldn't be correct to skip the `::regclass`, or would it?
Like 
```sql
SELECT run_command_on_placements('users_table',
...
```
Whats the difference between using the `::regclass` and not using it?
Just a guess: If I don't cast then the command would get executed multiple times on a worker for each replica of the `users_table` shard? And by casting to ``::regclass` it would be executed just one single time? Is this correct?
Thanks!

**Comment by mtuncer** [comment]
@mkurz  it is not related to ```::regclass```.  ```run_command_on_XXX``` functions are executed at coordinator. They prepare worker commands to be executed and forward them to workers.

In this particular example, ```run_command_on_placements``` gets the  function creation command, replaces ```%s``` with appropriate shard table name, and sends it to worker that contains this shard placement. 

To your question, ```::regclass``` casting is done only once.

**Comment by mkurz** [comment]
@mtuncer 
> `run_command_on_XXX` functions are executed at coordinator. They prepare worker commands to be executed and forward them to workers.

I am fully aware of this.

> To your question, ::regclass casting is done only once.

Of course, because that is done on the coordinator.

> In this particular example, `run_command_on_placements` gets the function creation command, replaces `%s` with appropriate shard table name, and sends it to worker that contains this shard placement.

I do understand this.
However, for `run_command_on_placements`,  [the docs say](https://docs.citusdata.com/en/v7.3/develop/reference_propagation.html#running-on-all-placements):
> The most granular level of execution is running a command across all shards **and their replicas** (aka placements). It can be useful for running data modification commands, which **must apply to every replica** to ensure consistency.

So if there is a replica of the shard table on the worker, does this mean that the function creation command will be executed twice (once for the shard and once for the shard replica)? That is what I don't understand.

**Comment by mtuncer** [comment]
> So if there is a replica of the shard table on the worker, does this mean that the function creation command will be executed twice (once for the shard and once for the shard replica)? That is what I don't understand.

That is correct, the command is executed on all placements of all shards for the given table.

**Comment by mkurz** [comment]
OK I think what I don't understand and needs clarification is:
In Citus terms, **what exactly is a replica of a shard and where will such replicas be saved**? For me a shard is a shard *table* in a *database* of the worker *instance* with a name like `my_table_102137`. For "reference tables" only *one* of such a shard table gets created on in a worker *database*, for distributed tables *multiple* of such `my_table_xxx` tables will be greated. Now:
* When will a replica of a shard *table* be created?
* Where will such a replica table be saved? Is it possible that a replica of a shard table is saved within the same worker DATABASE where the shard table itself is saved?
* Or is a replica of a shard (table) always saved in **another** database (so basically a replica of the whole database exists and within this replica database there are these replica shards)?

**Comment by onderkalaci** [comment]
This link might give you some insights: https://docs.citusdata.com/en/v7.3/get_started/concepts.html#shards  (e.g., shard vs shard placement)

**Comment by mkurz** [comment]
@onderkalaci thanks I read that now. From your link: 
> ...There are two replication “modes:” Citus replication and streaming replication. The former creates extra backup shard placements and runs queries against all of them that update any of them. The latter is more efficient and utilizes PostgreSQL’s streaming replication to back up the entire database of each node to a follower database. This is transparent and does not require the involvement of Citus metadata tables.

I also found more details about these two modes [here](https://docs.citusdata.com/en/v7.3/admin_guide/cluster_management.html#worker-node-failures).

Do I understand right that in case I configure Citus in the "PostgreSQL streaming replication" mode or just use Citus Cloud (which also uses streaming replication) I wouldn't even need to use `run_command_on_placements` anymore but `run_command_on_shards` would be enough? Because if I never ever use Citus second mode, "Citus shard replication", there wouldn't be any shard replicas managed by Citus itself to which I would want to propogate my commands - but instead that would be done automatically by PostgreSQL 10 streaming replication already? Is this correct?
Thanks for your time! (I want to understand how Citus works because I think about to use it for our new project and even consider Citus Cloud for it).

**Comment by onderkalaci** [comment]
>  I wouldn't even need to use run_command_on_placements anymore but run_command_on_shards would be enough

That's valid for [distributed tables](https://docs.citusdata.com/en/v7.3/get_started/concepts.html#type-1-distributed-tables). Still I personally follow the approach below:

In general the more common use case for `run_command_on_shards()` for queries that are always OK to run on a  single replica, such as `SELECT count(*) %s`.

Whereas, `run_command_on_placements()` is more often used for queries that need to be executed on all replicas of the shards such as `CREATE INDEX %s` where a modification is done on the placements.

Note that [reference tables](https://docs.citusdata.com/en/v7.3/get_started/concepts.html#type-2-reference-tables) are replicated to all primary nodes with Citus replication mode. So, you might still need to use `run_command_on_placements()` for reference tables if they involve any modifications on all of the shard placements.

> Because if I never ever use Citus second mode, "Citus shard replication", there wouldn't be any shard replicas managed by Citus itself to which I would want to propogate my commands

As I mentioned above, reference tables are exception. They are always replicated using Citus replication model on the cluster. You might still want to use `run_command_on_placements()`.

>  (I want to understand how Citus works because I think about to use it for our new project and even consider Citus Cloud for it).

Happy to help more if you need.  Feel free to ask any questions on our [slack channel](citus-public.slack.com) which is more active than github :) .

**Comment by mkurz** [comment]
@onderkalaci Very interesting! Thanks a lot, I think I understand now.

I also tested this now (with the docker images you provide). If I have five workers and I created a **reference** table and then run a simple insert like:
```sql
SELECT run_command_on_shards('my_reference_table', $cmd$INSERT INTO %s VALUES('foo');$cmd$);
```
the `foo` entry was inserted only in **one** of the workers `my_table` shards. If I use `run_command_on_placements` however it was inserted in **all** the workers `my_table` shards. This brings me to the question: It looks like for a reference table one of its shards is handled as the "primary shard" (the one that gets updated when using `run_command_on_shards`) whereas all others (in my case the 4 other shards on the 4 other workers) are handled as replicas of that first "primary shard". Is that correct?

As a conclusion of my learnings if I look at [this blog post](https://www.citusdata.com/blog/2018/04/04/raw-sql-access-with-row-level-security/) where you use `run_command_on_shards` everywhere is my assumption correct that in case someone would configure Citus with "Citus shard replication" (or by using Postgres < 10) then the blog post should instead use `run_command_on_placements`? (Seems like the blog post assumes that people use latest Postgres 10 where streaming replication is the default if not switched to the "old" citus mode by hand).

My last question: In your [first comment](https://github.com/citusdata/citus/issues/2126#issuecomment-384170125) in this thread  why did you use the cast `'users_table_ref'::regclass`? What is the advantage instead of just using the string `users_table`?

**Comment by onderkalaci** [comment]
Firstly, you should not rely on `run_command_*()` functions for production workloads such as `INSERT`s. These tools are only intended for utility/monitoring purposes. There are some `DDL` commands as in the blog post you've shared or other commands such as `CREATE FUNCTION` etc which are not propagated to the worker nodes by Citus automatically (e.g., mostly because we have not implemented propagating those commands yet).

> case the 4 other shards on the 4 other workers) are handled as replicas of that first "primary shard". Is that correct?

Actually not. All placements are equivalent from the database's perspective. If you run the same query on the coordinator, the query is executed using `2PC`, which means that a transaction is only committed when all the transactions on the shards are committed. Say you terminate one of the worker nodes, the query would fail. So, the shard placements of reference tables never diverge.

`run_command_on_shards` simply pick a random (or the first) shard placement and execute the command.

>  "Citus shard replication" (or by using Postgres < 10) then the blog post should instead use run_command_on_placements? (Seems like the blog post assumes that people use latest Postgres 10 where streaming replication is the default if not switched to the "old" citus mode by hand).

That's true. But note that Citus supports streaming replication for older Postgres versions as well. So, its not specific to PostgreSQL 10. 

> My last question: In your first comment in this thread why did you use the cast 'users_table_ref'::regclass? What is the advantage instead of just using the string users_table?

Oh, sorry for the confusion. That's not needed at all, I have the habit of casting types explicitly for `regclass`, which is not needed for sure since PostgreSQL already casts it to the correct type.

**Comment by mkurz** [comment]
> Firstly, you should not rely on `run_command_*()` functions for production workloads such as `INSERT`s. 

I know, I know, this was just a very stupid example to see the effects of the commands :wink: 

> `run_command_on_shards` simply pick a random (or the first) shard placement and execute the command.

Thanks, that is what I wanted to know.

Thank you very very much, now everything is clear to me!
I will use the slack channel for any other questions now :smile: 
Again, thank you for your time!

**Comment by begriffs** [comment]
This is an interesting discussion, and gives me some ideas for improving the documentation.

**Comment by thinhnp-it** [comment]
I'm a newbie in Citus and get same issue with @mkurz. Thanks all very much for an interesting discussion, guys.
